trieste/acquisition/sampler.py

```diff
@@ -0,0 +1,273 @@
+# Copyright 2020 The Trieste Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module is the home of the sampling functionality required by Trieste's
+acquisiiton functions.
+"""
+
+from __future__ import annotations
+
+from abc import ABC, abstractmethod
+
+import tensorflow as tf
+import tensorflow_probability as tfp
+from scipy.optimize import bisect
+
+from ..models import ProbabilisticModel
+from ..type import TensorType
+from ..utils import DEFAULTS
+
+
+class Sampler(ABC):
+    r"""
+    An :class:`Sampler` samples a specific quantity according to an
+    underlying :class:`ProbabilisticModel`.
+    """
+
+    def __init__(self, sample_size: int, model: ProbabilisticModel):
+        """
+        :param sample_size: The desired number of samples.
+        :param model: The model to sample from.
+        :raise ValueError (or InvalidArgumentError): If ``sample_size`` is not positive.
+        """
+        tf.debugging.assert_positive(sample_size)
+
+        self._sample_size = sample_size
+        self._model = model
+
+    def __repr__(self) -> str:
+        """"""
+        return f"{self.__class__.__name__}({self._sample_size!r}, {self._model!r})"
+
+    @abstractmethod
+    def sample(self, at: TensorType) -> TensorType:
+        """
+        :param at: Input points that define the sampler.
+        :return: Samples.
+        """
+
+
+class DiscreteThompsonSampler(Sampler):
```

Comment:
does it make sense to have this class? maybe we should take this opportunity to rewrite the TS code using `BatchReparametrizationSampler`? On the other hand this would limit its use to gpflow models.